zstash/extract.py

```diff
@@ -432,13 +441,47 @@ def extractFiles(  # noqa: C901
             if multiprocess_worker:
                 multiprocess_worker.set_curr_tar(files_row.tar)
 
-            if not os.path.exists(tfname):
-                # Will need to retrieve from HPSS
-                if config.hpss is not None:
-                    hpss: str = config.hpss
-                else:
-                    raise TypeError("Invalid config.hpss={}".format(config.hpss))
-                hpss_get(hpss, tfname, cache)
+            if config.hpss is not None:
+                hpss: str = config.hpss
+            else:
+                raise TypeError("Invalid config.hpss={}".format(config.hpss))
+            tries = args.retries + 1
+            # Set to True to test the `--retries` option with a forced failure.
+            # Then run `python -m unittest tests.test_extract.TestExtract.testExtractRetries`
+            test_retry = False
+            while tries > 0:
+                tries -= 1
+                try:
+                    if test_retry:
+                        test_retry = False
+                        raise RuntimeError
+                    if not os.path.exists(tfname):
+                        # Will need to retrieve from HPSS
+                        hpss_get(hpss, tfname, cache)
+                    elif cur and tars_table_exists(cur):
+                        logger.info(
+                            f"{tfname} exists. Checking expected size matches actual size."
+                        )
+                        actual_size = os.path.getsize(tfname)
```

Comment:
I did a test by doing the following:

Changed this code block to be:
```
                    if not os.path.exists(tfname):
                        # Will need to retrieve from HPSS
                        hpss_get(hpss, tfname, cache)
                        raise RuntimeError # NEW -- on retry, tfname will exist and logic will go into elif block
                    elif cur and tars_table_exists(cur):
                        logger.info(
                            f"{tfname} exists. Checking expected size matches actual size."
                        )
                        actual_size = os.path.getsize("/global/homes/f/forsyth/zstash_test/add_files.sh") # CHANGED -- this is a much smaller file, so the sizes won't match
```
and run a modified version of the tutorial code:
```
$ emacs setup.sh
mkdir zstash_demo
mkdir zstash_demo/empty_dir
mkdir zstash_demo/dir
echo 'file0 stuff' > zstash_demo/file0.txt
echo '' > zstash_demo/file_empty.txt
echo 'file1 stuff' > zstash_demo/dir/file1.txt
$ emacs add_files.sh
mkdir zstash_demo/dir2
echo 'file2 stuff' > zstash_demo/dir2/file2.txt
echo 'file1 stuff with changes' > zstash_demo/dir/file1.txt
$ chmod 700 setup.sh
$ chmod 700 add_files.sh

$ ./setup.sh
$ zstash create --hpss=zstash_retries zstash_demo
$ ./add_files.sh
$ cd zstash_demo
$ zstash update --hpss=zstash_retries
$ cd ..
$ mkdir zstash_extraction2 && cd zstash_extraction2
$ zstash extract --hpss=zstash_retries --retries=3
```